File: desktop/FedMes.Desktop.Core/AccountSecurity/AccountVaultCrypto.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;

namespace FedMes.Desktop.Core.AccountSecurity;

public static class AccountVaultCrypto
{
    public static byte[] GenerateAccountRootKey() => RandomNumberGenerator.GetBytes(AccountSecurityProtocol.AccountRootKeyBytes);

    public static string GenerateRecoveryKey() => Base64UrlEncode(RandomNumberGenerator.GetBytes(AccountSecurityProtocol.RecoveryKeyBytes));

    public static EncryptedAccountVault EncryptVault(string serverUrl, string username, ReadOnlySpan<byte> accountRootKey, long previousRevision)
    {
        ValidateRootKey(accountRootKey);
        long revision = checked(previousRevision + 1);
        byte[] testRecord = RandomNumberGenerator.GetBytes(32);
        byte[] plaintext = JsonSerializer.SerializeToUtf8Bytes(new
        {
            version = AccountSecurityProtocol.VaultVersion,
            crypto_version = AccountSecurityProtocol.CryptoVersion,
            username,
            revision,
            test_record = Convert.ToBase64String(testRecord),
            room_keys = new { },
            file_keys = new { },
            rotation_state = new { },
        });
        byte[] nonce = RandomNumberGenerator.GetBytes(AccountSecurityProtocol.NonceBytes);
        byte[] ciphertext = new byte[plaintext.Length];
        byte[] tag = new byte[16];
        byte[] aad = VaultAad(serverUrl, username, revision);
        try
        {
            using var aes = new AesGcm(accountRootKey, tag.Length);
            aes.Encrypt(nonce, plaintext, ciphertext, tag, aad);
            byte[] combined = new byte[ciphertext.Length + tag.Length];
            ciphertext.CopyTo(combined, 0);
            tag.CopyTo(combined, ciphertext.Length);
            return new EncryptedAccountVault(
                revision,
                AccountSecurityProtocol.VaultVersion,
                AccountSecurityProtocol.CryptoVersion,
                AccountSecurityProtocol.AadVersion,
                nonce,
                combined,
                Convert.ToHexStringLower(SHA256.HashData(combined)));
        }
        finally
        {
            CryptographicOperations.ZeroMemory(testRecord);
            CryptographicOperations.ZeroMemory(plaintext);
            CryptographicOperations.ZeroMemory(ciphertext);
            CryptographicOperations.ZeroMemory(tag);
            CryptographicOperations.ZeroMemory(aad);
        }
    }

    public static void VerifyVault(string serverUrl, string username, ReadOnlySpan<byte> accountRootKey, EncryptedAccountVault vault)
    {
        ValidateRootKey(accountRootKey);
        if (vault.VaultVersion != AccountSecurityProtocol.VaultVersion ||
            vault.CryptoVersion != AccountSecurityProtocol.CryptoVersion ||
            vault.AadVersion != AccountSecurityProtocol.AadVersion ||
            vault.Revision <= 0 || vault.Ciphertext.Length < 17 ||
            vault.Nonce.Length != AccountSecurityProtocol.NonceBytes)
        {
            throw new CryptographicException("Encrypted vault metadata is invalid.");
        }
        byte[] computedHash = SHA256.HashData(vault.Ciphertext);
        byte[] expectedHash = ParseSha256(vault.CiphertextSha256);
        try
        {
            if (!CryptographicOperations.FixedTimeEquals(computedHash, expectedHash))
            {
                throw new CryptographicException("Encrypted vault hash mismatch.");
            }
        }
        finally
        {
            CryptographicOperations.ZeroMemory(computedHash);
            CryptographicOperations.ZeroMemory(expectedHash);
        }
        int contentLength = vault.Ciphertext.Length - 16;
        byte[] plaintext = new byte[contentLength];
        byte[] aad = VaultAad(serverUrl, username, vault.Revision);
        try
        {
            using var aes = new AesGcm(accountRootKey, 16);
            aes.Decrypt(vault.Nonce, vault.Ciphertext.AsSpan(0, contentLength), vault.Ciphertext.AsSpan(contentLength), plaintext, aad);
            using JsonDocument document = JsonDocument.Parse(plaintext);
            JsonElement root = document.RootElement;
            byte[] testRecord = Convert.FromBase64String(root.GetProperty("test_record").GetString() ?? string.Empty);
            try
            {
                if (root.GetProperty("version").GetInt32() != vault.VaultVersion ||
                    root.GetProperty("crypto_version").GetInt32() != vault.CryptoVersion ||
                    !string.Equals(root.GetProperty("username").GetString(), username, StringComparison.Ordinal) ||
                    root.GetProperty("revision").GetInt64() != vault.Revision ||
                    testRecord.Length != 32)
                {
                    throw new CryptographicException("Encrypted vault test record failed.");
                }
            }
            finally
            {
                CryptographicOperations.ZeroMemory(testRecord);
            }
        }
        finally
        {
            CryptographicOperations.ZeroMemory(plaintext);
            CryptographicOperations.ZeroMemory(aad);
        }
    }

    public static EncryptedRecoveryPackage CreateRecoveryPackage(
        string serverUrl,
        string username,
        string recoveryKeyText,
        ReadOnlySpan<byte> accountRootKey,
        long vaultRevision)
    {
        ValidateRootKey(accountRootKey);
        byte[] recoveryKey = ParseRecoveryKey(recoveryKeyText);
        byte[] salt = RandomNumberGenerator.GetBytes(AccountSecurityProtocol.RecoverySaltBytes);
        byte[] wrappingKey = DeriveRecoveryWrappingKey(recoveryKey, salt);
        string id = Guid.NewGuid().ToString();
        byte[] nonce = RandomNumberGenerator.GetBytes(AccountSecurityProtocol.NonceBytes);
        byte[] encrypted = new byte[AccountSecurityProtocol.AccountRootKeyBytes];
        byte[] tag = new byte[16];
        byte[] aad = RecoveryAad(serverUrl, username, id, vaultRevision);
        try
        {
            using var aes = new AesGcm(wrappingKey, tag.Length);
            aes.Encrypt(nonce, accountRootKey, encrypted, tag, aad);
            byte[] combined = new byte[encrypted.Length + tag.Length];
            encrypted.CopyTo(combined, 0);
            tag.CopyTo(combined, encrypted.Length);
            return new EncryptedRecoveryPackage(
                id,
                vaultRevision,
                AccountSecurityProtocol.RecoveryPackageVersion,
                AccountSecurityProtocol.CryptoVersion,
                AccountSecurityProtocol.AadVersion,
                "PBKDF2-HMAC-SHA256",
                JsonSerializer.Serialize(new
                {
                    iterations = AccountSecurityProtocol.RecoveryKdfIterations,
                    key_bits = 256,
                    password_encoding = "base64url-canonical",
                }),
                salt,
                nonce,
                combined,
                Convert.ToHexStringLower(SHA256.HashData(combined)));
        }
        finally
        {
            CryptographicOperations.ZeroMemory(recoveryKey);
            CryptographicOperations.ZeroMemory(wrappingKey);
            CryptographicOperations.ZeroMemory(encrypted);
            CryptographicOperations.ZeroMemory(tag);
            CryptographicOperations.ZeroMemory(aad);
        }
    }

    public static byte[] CreateAccessVerifier(string serverUrl, string username, ReadOnlySpan<byte> accountRootKey, long vaultRevision)
    {
        ValidateRootKey(accountRootKey);
        ArgumentOutOfRangeException.ThrowIfNegativeOrZero(vaultRevision);
        byte[] context = Encoding.UTF8.GetBytes($"fedmes-vault-access-v1\n{serverUrl}\n{username}\n{vaultRevision}");
        try
        {
            return HMACSHA256.HashData(accountRootKey, context);
        }
        finally
        {
            CryptographicOperations.ZeroMemory(context);
        }
    }

    public static byte[] RecoverAccountRootKey(string serverUrl, string username, string recoveryKeyText, EncryptedRecoveryPackage package)
    {
        ValidateRecoveryPackage(package);
        byte[] recoveryKey = ParseRecoveryKey(recoveryKeyText);
        byte[] wrappingKey = DeriveRecoveryWrappingKey(recoveryKey, package.Salt);
        byte[] plaintext = new byte[AccountSecurityProtocol.AccountRootKeyBytes];
        byte[] aad = RecoveryAad(serverUrl, username, package.Id, package.VaultRevision);
        byte[] computedHash = SHA256.HashData(package.Ciphertext);
        byte[] expectedHash = ParseSha256(package.CiphertextSha256);
        try
        {
            if (!CryptographicOperations.FixedTimeEquals(computedHash, expectedHash))
            {
                throw new CryptographicException("Recovery package is invalid.");
            }
            using var aes = new AesGcm(wrappingKey, 16);
            aes.Decrypt(package.Nonce, package.Ciphertext.AsSpan(0, AccountSecurityProtocol.AccountRootKeyBytes), package.Ciphertext.AsSpan(AccountSecurityProtocol.AccountRootKeyBytes), plaintext, aad);
            return plaintext;
        }
        catch (CryptographicException ex)
        {
            // Normalize all recovery-authentication failures to the public base exception.
            // This keeps the API stable across .NET runtime versions where AesGcm may throw
            // a more specific AuthenticationTagMismatchException.
            CryptographicOperations.ZeroMemory(plaintext);
            throw new CryptographicException("Recovery package authentication failed.", ex);
        }
        catch
        {
            CryptographicOperations.ZeroMemory(plaintext);
            throw;
        }
        finally
        {
            CryptographicOperations.ZeroMemory(recoveryKey);
            CryptographicOperations.ZeroMemory(wrappingKey);
            CryptographicOperations.ZeroMemory(aad);
            CryptographicOperations.ZeroMemory(computedHash);
            CryptographicOperations.ZeroMemory(expectedHash);
        }
    }

    private static void ValidateRecoveryPackage(EncryptedRecoveryPackage package)
    {
        ArgumentNullException.ThrowIfNull(package);
        if (!string.Equals(package.KdfName, "PBKDF2-HMAC-SHA256", StringComparison.Ordinal) ||
            package.PackageVersion != AccountSecurityProtocol.RecoveryPackageVersion ||
            package.CryptoVersion != AccountSecurityProtocol.CryptoVersion ||
            package.AadVersion != AccountSecurityProtocol.AadVersion || package.VaultRevision <= 0 ||
            package.Salt.Length != AccountSecurityProtocol.RecoverySaltBytes ||
            package.Nonce.Length != AccountSecurityProtocol.NonceBytes ||
            package.Ciphertext.Length != AccountSecurityProtocol.AccountRootKeyBytes + 16)
        {
            throw new CryptographicException("Recovery package metadata is invalid.");
        }
        using JsonDocument parameters = JsonDocument.Parse(package.KdfParameters);
        JsonElement root = parameters.RootElement;
        if (root.GetProperty("iterations").GetInt32() != AccountSecurityProtocol.RecoveryKdfIterations ||
            root.GetProperty("key_bits").GetInt32() != 256 ||
            !string.Equals(root.GetProperty("password_encoding").GetString(), "base64url-canonical", StringComparison.Ordinal))
        {
            throw new CryptographicException("Recovery package KDF parameters are invalid.");
        }
    }

    private static byte[] ParseSha256(string value)
    {
        if (string.IsNullOrWhiteSpace(value) || value.Length != 64 || !value.All(Uri.IsHexDigit))
        {
            throw new CryptographicException("SHA-256 value is invalid.");
        }
        return Convert.FromHexString(value);
    }

    private static byte[] ParseRecoveryKey(string text)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(text);
        string normalized = string.Concat(text.Where(c => !char.IsWhiteSpace(c)));
        byte[] key = Base64UrlDecode(normalized);
        if (key.Length != AccountSecurityProtocol.RecoveryKeyBytes)
        {
            CryptographicOperations.ZeroMemory(key);
            throw new CryptographicException("Recovery Key has an invalid size.");
        }
        return key;
    }

    private static void ValidateRootKey(ReadOnlySpan<byte> value)
    {
        if (value.Length != AccountSecurityProtocol.AccountRootKeyBytes) throw new CryptographicException("Account Root Key has an invalid size.");
    }

    private static byte[] VaultAad(string serverUrl, string username, long revision) =>
        Encoding.UTF8.GetBytes(
            $"fedmes-vault\nprotocol={AccountSecurityProtocol.Version}\nserver={serverUrl}\naccount={username}\n" +
            $"vault_version={AccountSecurityProtocol.VaultVersion}\nrevision={revision}\n" +
            $"crypto_version={AccountSecurityProtocol.CryptoVersion}\naad_version={AccountSecurityProtocol.AadVersion}");

    private static byte[] RecoveryAad(string serverUrl, string username, string id, long revision) =>
        Encoding.UTF8.GetBytes(
            $"fedmes-recovery\nprotocol={AccountSecurityProtocol.Version}\nserver={serverUrl}\naccount={username}\n" +
            $"package_id={id}\nvault_version={AccountSecurityProtocol.VaultVersion}\nvault_revision={revision}\n" +
            $"crypto_version={AccountSecurityProtocol.CryptoVersion}\naad_version={AccountSecurityProtocol.AadVersion}");

    private static byte[] DeriveRecoveryWrappingKey(byte[] recoveryKey, byte[] salt)
    {
        string canonical = Convert.ToBase64String(recoveryKey).TrimEnd('=').Replace('+', '-').Replace('/', '_');
        byte[] password = Encoding.ASCII.GetBytes(canonical);
        try
        {
            return Rfc2898DeriveBytes.Pbkdf2(
                password,
                salt,
                AccountSecurityProtocol.RecoveryKdfIterations,
                HashAlgorithmName.SHA256,
                32);
        }
        finally
        {
            CryptographicOperations.ZeroMemory(password);
        }
    }

    private static string Base64UrlEncode(byte[] bytes)
    {
        try { return Convert.ToBase64String(bytes).TrimEnd('=').Replace('+', '-').Replace('/', '_'); }
        finally { CryptographicOperations.ZeroMemory(bytes); }
    }

    private static byte[] Base64UrlDecode(string text) => Convert.FromBase64String(text.Replace('-', '+').Replace('_', '/') + new string('=', (4 - text.Length % 4) % 4));
}

File: android/app/src/main/java/com/fedmes/app/accountsecurity/AccountVaultCrypto.kt
package com.fedmes.app.accountsecurity

import org.json.JSONObject
import java.util.Base64
import java.security.MessageDigest
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.Mac
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.PBEKeySpec
import javax.crypto.spec.SecretKeySpec

class AccountVaultCrypto(
    private val secureRandom: SecureRandom = SecureRandom(),
) {
    fun generateAccountRootKey(): ByteArray = ByteArray(AccountSecurityProtocol.ACCOUNT_ROOT_KEY_BYTES).also {
        secureRandom.nextBytes(it)
    }

    fun generateRecoveryKeyText(): String {
        val bytes = ByteArray(AccountSecurityProtocol.RECOVERY_KEY_BYTES)
        return try {
            secureRandom.nextBytes(bytes)
            Base64.getUrlEncoder().withoutPadding().encodeToString(bytes)
        } finally {
            bytes.fill(0)
        }
    }

    fun encryptVault(
        serverUrl: String,
        username: String,
        accountRootKey: ByteArray,
        previousRevision: Long,
        legacyKeyMaterial: ByteArray? = null,
    ): Pair<EncryptedAccountVault, Long> {
        require(accountRootKey.size == AccountSecurityProtocol.ACCOUNT_ROOT_KEY_BYTES)
        val revision = previousRevision + 1L
        val testRecord = ByteArray(32)
        secureRandom.nextBytes(testRecord)
        val payload = JSONObject()
            .put("version", AccountSecurityProtocol.VAULT_VERSION)
            .put("crypto_version", AccountSecurityProtocol.CRYPTO_VERSION)
            .put("username", username)
            .put("revision", revision)
            .put("test_record", Base64.getEncoder().encodeToString(testRecord))
            .put("room_keys", JSONObject())
            .put("file_keys", JSONObject())
            .put("rotation_state", JSONObject())
            .put(
                "legacy_fmk",
                legacyKeyMaterial?.let { Base64.getEncoder().encodeToString(it) } ?: JSONObject.NULL,
            )
            .toString()
            .toByteArray(Charsets.UTF_8)
        val nonce = ByteArray(AccountSecurityProtocol.NONCE_BYTES)
        secureRandom.nextBytes(nonce)
        val aad = vaultAad(serverUrl, username, revision)
        val ciphertext = try {
            aesGcmEncrypt(accountRootKey, nonce, aad, payload)
        } finally {
            payload.fill(0)
            testRecord.fill(0)
            aad.fill(0)
        }
        val hash = sha256Hex(ciphertext)
        return EncryptedAccountVault(
            revision = revision,
            vaultVersion = AccountSecurityProtocol.VAULT_VERSION,
            cryptoVersion = AccountSecurityProtocol.CRYPTO_VERSION,
            aadVersion = AccountSecurityProtocol.AAD_VERSION,
            nonce = nonce,
            ciphertext = ciphertext,
            ciphertextSha256Hex = hash,
        ) to revision
    }

    fun verifyAndDecryptVault(
        serverUrl: String,
        username: String,
        accountRootKey: ByteArray,
        vault: EncryptedAccountVault,
    ): JSONObject {
        require(accountRootKey.size == AccountSecurityProtocol.ACCOUNT_ROOT_KEY_BYTES)
        if (vault.vaultVersion != AccountSecurityProtocol.VAULT_VERSION ||
            vault.cryptoVersion != AccountSecurityProtocol.CRYPTO_VERSION ||
            vault.aadVersion != AccountSecurityProtocol.AAD_VERSION ||
            vault.revision <= 0L ||
            vault.nonce.size != AccountSecurityProtocol.NONCE_BYTES ||
            vault.ciphertext.size < 17
        ) {
            throw AccountSecurityException("vault_metadata_invalid")
        }
        if (!constantTimeHexEquals(vault.ciphertextSha256Hex, sha256Hex(vault.ciphertext))) {
            throw AccountSecurityException("vault_hash_mismatch")
        }
        val aad = vaultAad(serverUrl, username, vault.revision)
        val plaintext = try {
            aesGcmDecrypt(accountRootKey, vault.nonce, aad, vault.ciphertext)
        } catch (error: Exception) {
            throw AccountSecurityException("vault_decryption_failed", error)
        } finally {
            aad.fill(0)
        }
        return try {
            val root = JSONObject(plaintext.toString(Charsets.UTF_8))
            val testRecord = Base64.getDecoder().decode(root.getString("test_record"))
            try {
                if (root.getInt("version") != vault.vaultVersion ||
                    root.getInt("crypto_version") != vault.cryptoVersion ||
                    root.getString("username") != username ||
                    root.getLong("revision") != vault.revision ||
                    testRecord.size != 32
                ) {
                    throw AccountSecurityException("vault_test_record_invalid")
                }
            } finally {
                testRecord.fill(0)
            }
            root
        } finally {
            plaintext.fill(0)
        }
    }

    fun createRecoveryPackage(
        serverUrl: String,
        username: String,
        recoveryKeyText: String,
        accountRootKey: ByteArray,
        vaultRevision: Long,
    ): EncryptedRecoveryPackage {
        val recoveryKey = parseRecoveryKey(recoveryKeyText)
        val salt = ByteArray(AccountSecurityProtocol.RECOVERY_SALT_BYTES)
        val nonce = ByteArray(AccountSecurityProtocol.NONCE_BYTES)
        secureRandom.nextBytes(salt)
        secureRandom.nextBytes(nonce)
        val wrappingKey = deriveRecoveryWrappingKey(recoveryKey, salt)
        val packageId = AccountSecurityProtocol.newRequestId()
        val aad = recoveryAad(serverUrl, username, packageId, vaultRevision)
        val ciphertext = try {
            aesGcmEncrypt(wrappingKey, nonce, aad, accountRootKey)
        } finally {
            recoveryKey.fill(0)
            wrappingKey.fill(0)
            aad.fill(0)
        }
        return EncryptedRecoveryPackage(
            id = packageId,
            vaultRevision = vaultRevision,
            packageVersion = AccountSecurityProtocol.RECOVERY_PACKAGE_VERSION,
            cryptoVersion = AccountSecurityProtocol.CRYPTO_VERSION,
            aadVersion = AccountSecurityProtocol.AAD_VERSION,
            kdfName = AccountSecurityProtocol.RECOVERY_KDF_NAME,
            kdfParameters = JSONObject()
                .put("iterations", AccountSecurityProtocol.RECOVERY_KDF_ITERATIONS)
                .put("key_bits", 256)
                .put("password_encoding", "base64url-canonical")
                .toString(),
            salt = salt,
            nonce = nonce,
            ciphertext = ciphertext,
            ciphertextSha256Hex = sha256Hex(ciphertext),
        )
    }

    fun recoverAccountRootKey(
        serverUrl: String,
        username: String,
        recoveryKeyText: String,
        recoveryPackage: EncryptedRecoveryPackage,
    ): ByteArray {
        val kdfParameters = runCatching { JSONObject(recoveryPackage.kdfParameters) }
            .getOrElse { throw AccountSecurityException("recovery_package_invalid", it) }
        if (recoveryPackage.kdfName != AccountSecurityProtocol.RECOVERY_KDF_NAME ||
            recoveryPackage.packageVersion != AccountSecurityProtocol.RECOVERY_PACKAGE_VERSION ||
            recoveryPackage.cryptoVersion != AccountSecurityProtocol.CRYPTO_VERSION ||
            recoveryPackage.aadVersion != AccountSecurityProtocol.AAD_VERSION ||
            recoveryPackage.vaultRevision <= 0L ||
            recoveryPackage.salt.size != AccountSecurityProtocol.RECOVERY_SALT_BYTES ||
            recoveryPackage.nonce.size != AccountSecurityProtocol.NONCE_BYTES ||
            recoveryPackage.ciphertext.size != AccountSecurityProtocol.ACCOUNT_ROOT_KEY_BYTES + 16 ||
            kdfParameters.optInt("iterations") != AccountSecurityProtocol.RECOVERY_KDF_ITERATIONS ||
            kdfParameters.optInt("key_bits") != 256 ||
            kdfParameters.optString("password_encoding") != "base64url-canonical" ||
            !constantTimeHexEquals(recoveryPackage.ciphertextSha256Hex, sha256Hex(recoveryPackage.ciphertext))
        ) {
            throw AccountSecurityException("recovery_package_invalid")
        }
        val recoveryKey = parseRecoveryKey(recoveryKeyText)
        val wrappingKey = deriveRecoveryWrappingKey(recoveryKey, recoveryPackage.salt)
        val aad = recoveryAad(
            serverUrl,
            username,
            recoveryPackage.id,
            recoveryPackage.vaultRevision,
        )
        return try {
            aesGcmDecrypt(wrappingKey, recoveryPackage.nonce, aad, recoveryPackage.ciphertext).also {
                if (it.size != AccountSecurityProtocol.ACCOUNT_ROOT_KEY_BYTES) {
                    it.fill(0)
                    throw AccountSecurityException("recovery_key_invalid")
                }
            }
        } catch (error: AccountSecurityException) {
            throw error
        } catch (error: Exception) {
            throw AccountSecurityException("recovery_key_invalid", error)
        } finally {
            recoveryKey.fill(0)
            wrappingKey.fill(0)
            aad.fill(0)
        }
    }

    fun accessVerifier(
        serverUrl: String,
        username: String,
        accountRootKey: ByteArray,
        vaultRevision: Long,
    ): ByteArray {
        require(accountRootKey.size == AccountSecurityProtocol.ACCOUNT_ROOT_KEY_BYTES)
        require(vaultRevision > 0L)
        val context = "fedmes-vault-access-v1\n$serverUrl\n$username\n$vaultRevision"
            .toByteArray(Charsets.UTF_8)
        return try {
            val mac = Mac.getInstance("HmacSHA256")
            mac.init(SecretKeySpec(accountRootKey, "HmacSHA256"))
            mac.doFinal(context)
        } finally {
            context.fill(0)
        }
    }

    fun parseRecoveryKey(text: String): ByteArray {
        val normalized = text.trim().replace(" ", "").replace("\n", "").replace("\r", "")
        val decoded = try {
            Base64.getUrlDecoder().decode(normalized)
        } catch (error: IllegalArgumentException) {
            throw AccountSecurityException("recovery_key_invalid", error)
        }
        if (decoded.size != AccountSecurityProtocol.RECOVERY_KEY_BYTES) {
            decoded.fill(0)
            throw AccountSecurityException("recovery_key_invalid")
        }
        return decoded
    }

    private fun deriveRecoveryWrappingKey(recoveryKey: ByteArray, salt: ByteArray): ByteArray {
        val passwordChars = Base64.getUrlEncoder().withoutPadding()
            .encodeToString(recoveryKey)
            .toCharArray()
        val spec = PBEKeySpec(
            passwordChars,
            salt,
            AccountSecurityProtocol.RECOVERY_KDF_ITERATIONS,
            256,
        )
        return try {
            SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256").generateSecret(spec).encoded
        } finally {
            passwordChars.fill('\u0000')
            spec.clearPassword()
        }
    }

    private fun aesGcmEncrypt(key: ByteArray, nonce: ByteArray, aad: ByteArray, plaintext: ByteArray): ByteArray {
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(128, nonce))
        cipher.updateAAD(aad)
        return cipher.doFinal(plaintext)
    }

    private fun aesGcmDecrypt(key: ByteArray, nonce: ByteArray, aad: ByteArray, ciphertext: ByteArray): ByteArray {
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(128, nonce))
        cipher.updateAAD(aad)
        return cipher.doFinal(ciphertext)
    }

    private fun vaultAad(serverUrl: String, username: String, revision: Long): ByteArray =
        listOf(
            "fedmes-vault",
            "protocol=${AccountSecurityProtocol.VERSION}",
            "server=$serverUrl",
            "account=$username",
            "vault_version=${AccountSecurityProtocol.VAULT_VERSION}",
            "revision=$revision",
            "crypto_version=${AccountSecurityProtocol.CRYPTO_VERSION}",
            "aad_version=${AccountSecurityProtocol.AAD_VERSION}",
        ).joinToString("\n").toByteArray(Charsets.UTF_8)

    private fun recoveryAad(
        serverUrl: String,
        username: String,
        packageId: String,
        vaultRevision: Long,
    ): ByteArray = listOf(
        "fedmes-recovery",
        "protocol=${AccountSecurityProtocol.VERSION}",
        "server=$serverUrl",
        "account=$username",
        "package_id=$packageId",
        "vault_version=${AccountSecurityProtocol.VAULT_VERSION}",
        "vault_revision=$vaultRevision",
        "crypto_version=${AccountSecurityProtocol.CRYPTO_VERSION}",
        "aad_version=${AccountSecurityProtocol.AAD_VERSION}",
    ).joinToString("\n").toByteArray(Charsets.UTF_8)

    private fun sha256Hex(value: ByteArray): String = MessageDigest.getInstance("SHA-256")
        .digest(value)
        .joinToString("") { byte -> "%02x".format(byte.toInt() and 0xff) }

    private fun constantTimeHexEquals(left: String, right: String): Boolean {
        val leftBytes = left.lowercase().toByteArray(Charsets.US_ASCII)
        val rightBytes = right.lowercase().toByteArray(Charsets.US_ASCII)
        return try {
            MessageDigest.isEqual(leftBytes, rightBytes)
        } finally {
            leftBytes.fill(0)
            rightBytes.fill(0)
        }
    }
}
